log/sockstatlog/logger.go

```diff
@@ -0,0 +1,124 @@
+// Package sockstatlog provides a logger for capturing and storing network socket stats.
+package sockstatlog
+
+import (
+	"encoding/json"
+	"path/filepath"
+	"time"
+
+	"tailscale.com/logtail/filch"
+	"tailscale.com/net/sockstats"
+)
+
+// pollPeriod specifies how often to poll for socket stats.
+const pollPeriod = time.Second / 10
+
+// Logger logs statistics about network sockets.
+type Logger struct {
+	ticker time.Ticker
+
+	// last is the set of socket stats we saw.
+	last     statMap
+	lastTime time.Time
+
+	logbuffer *filch.Filch
+}
+
+// statMap contains a map of stat values for labels.
+// In the net/sockstats package, a SockStat always represents total bytes transferred.
+// In this package, we reuse the type to also represent bytes transferred within a specific time interval.
+type statMap map[sockstats.Label]sockstats.SockStat
+
+type logEntry struct {
+	// Start and End of the time interval as Unix timestamps in milliseconds.
+	Start int64 `json:"start"`
+	End   int64 `json:"end"`
+
+	// Stats records the stats for the time interval.
+	Stats statMap `json:"stats"`
+}
+
+// NewLogger returns a new Logger that will store stats in logdir.
+// On platforms that do not support sockstat logging, a nil Logger will be returned.
+// The returned Logger is not yet running.
+func NewLogger(logdir string) (*Logger, error) {
+	if !loggingEnabled {
+		return nil, nil
+	}
+	filchPrefix := filepath.Join(logdir, "sockstats")
+	filch, err := filch.New(filchPrefix, filch.Options{ReplaceStderr: false})
+	if err != nil {
+		return nil, err
+	}
+
+	return &Logger{
+		ticker:    *time.NewTicker(pollPeriod),
+		logbuffer: filch,
+	}, nil
+}
+
+func (l *Logger) Start() {
+	go l.poll()
+}
+
+// poll fetches the current socket stats at the configured time interval,
+// calculates the delta since the last poll, and logs any non-zero values.
+// This method does not return.
+func (l *Logger) poll() {
+	enc := json.NewEncoder(l.logbuffer)
+	for t := range l.ticker.C {
+		stats := sockstats.Get()
+		diffstats, changed := delta(l.last, stats.Stats)
+		if changed {
+			entry := logEntry{
+				Start: l.lastTime.UnixMilli(),
+				End:   t.UnixMilli(),
+				Stats: diffstats,
+			}
+			enc.Encode(entry)
+		}
+		l.lastTime = t
+		l.last = stats.Stats
+	}
+}
+
+func (l *Logger) Shutdown() {
+	l.ticker.Stop()
+	l.logbuffer.Close()
+}
+
+// delta calculates the delta stats between two snapshots.
+// a and b are maps of SockStat values containing total bytes transferred.
+// The returned map contains the bytes transferred between the two snapshots.
+// Zero values are omitted from the returned map, and changed indicates if any stats have changed during the interval.
+func delta(a, b statMap) (diffstats statMap, changed bool) {
+	diffstats = make(statMap, len(b))
+	for label, bs := range b {
+		var labelChanged bool
+		as := a[label]
+		stat := sockstats.SockStat{
```

Comment:
I assume this always does an alloc to create the struct, even if we end up throwing it away because nothing changed.

May want to have a fast check of old/new TxBytes/RxBytes instead.

Reply:
done